docs/lessons/09-interitance.md

```diff
@@ -0,0 +1,107 @@
+# 09 - Inheritance & overriding
+<Teacher name="Alina"></Teacher>
+
+## Goals
+- Understand what is Inheritance and why do we need it 
+- Understand what is Overriding
+- Understand what is difference between Overriding and Overloading
+- Create Classes with hierarchy
+- Create overridden methods and constructors
+
+## Slides
+<GoogleSlides src="https://docs.google.com/presentation/d/13uaQDLuHeg8DmFKRZfFNGOjz3W4rMXszzcP0250WXP8/embed"></GoogleSlides>
+
+## Inheritance
+In object-oriented programming, inheritance is the mechanism of basing an object or class upon another object or class, retaining similar implementation. Also defined as deriving new classes from existing ones such as super class or base class and then forming them into a hierarchy of classes. 
+
+### Definitions
+A class that is derived from another class is called a subclass (also a derived class, extended class, or child class). 
+
+The class from which the subclass is derived is called a superclass (also a base class or a parent class).
+
+Excepting Object, which has no superclass, every class has one and only one direct superclass (single inheritance). In the absence of any other explicit superclass, every class is implicitly a subclass of Object.
+
+### Example
+
+```java
+import java.util.ArrayList;
+
+public class Person {
+
+  public String name;
+  public int age;
+  public String city;
+  public ArrayList<Person> friends;
+
+  public Person(String name, int age, String city) {
+    this.name = name;
+    this.age = age;
+    this.city = city;
+    this.friends = new ArrayList<>();
+  }
+
+  public void introduce() {
+    System.out.println("My name is " + this.name);
+    System.out.println("I'm " + this.age + " years old");
+    System.out.println("My friends are:");
+    for (Person s : friends) {
+      System.out.println(s.name);
+    }
+  }
+}
+
+```
+
+1. Let's implement together classes `Student` & `Teacher`, that are subclasses of the class `Person`
+1.1. `Student` has a list of subjects that they study
+1.2. `Teacher` has a subject that they teach, how many years of experience they have
+1.3. Add a method `study(String subject)` to the `Student` class, that prints
+1.4. Add a method `teach()` to the `Teacher` class
+2. Let's add constructors to the `Person`, `Student` & `Teacher` classes
+3. Let's override `introduce()` method of the `Student` and `Teacher` classes
+
+### What you can do in a Subclass
+
+A subclass inherits all of the public and protected members of its parent, no matter what package the subclass is in. If the subclass is in the same package as its parent, it also inherits the package-private members of the parent. You can use the inherited members as is, replace them, hide them, or supplement them with new members:
```

Comment:
Have we, in a previous class, introduced the notion of package to the students? Or will it be done in this class?

Reply:
good point, I will do it right before Access modifiers recap